File: update-deals-action.mjs
// update-deals-action.mjs
// Daily deal updater — asks Claude for today's best deals
// No external scraping needed — Claude knows current deals

import https from 'https';
import fs from 'fs';

const KEY = process.env.ANTHROPIC_API_KEY;

async function getDeals() {
  const today = new Date().toLocaleDateString('en-US', {
    weekday: 'long', year: 'numeric', month: 'long', day: 'numeric'
  });

  const prompt = `You are a deal curator for Scoop Deals (myscoopdeals.com). Today is ${today}.

Find the 12 best current deals, promo codes and sales available right now from stores like Amazon, Target, Walmart, Best Buy, Nike, Ulta, Chewy, Kohl's, Nordstrom, Wayfair, Home Depot, Lowe's, Sephora, Bath & Body Works, Old Navy, Gap, Lululemon, REI, Dick's Sporting Goods, CVS, Walgreens and other major retailers.

Return ONLY a valid JSON array of exactly 12 objects. No markdown, no explanation. Start directly with [

Each object must have exactly these fields:
{
  "brand": "Brand Name",
  "domain": "brand.com",
  "badge": "Hot Deal",
  "badgeStyle": "badge-hot",
  "category": "Category · Subcategory",
  "title": "Specific deal title with actual price or % savings",
  "description": "2-3 sentences explaining the deal, how to get it, any restrictions",
  "discount": "40% OFF",
  "originalPrice": "Was $X or empty string",
  "code": "PROMOCODE or empty string",
  "noCode": "✓ No code needed or empty string if code exists",
  "link": "https://direct-url-to-deal",
  "expiry": "⏰ Expires [date] or ⚡ Today only or empty string"
}

badge options: Hot Deal, New, Verified, Freebie, Flash Sale, Email Exclusive
badgeStyle options: badge-hot, badge-new, badge-email, badge-web, badge-free

Rules:
- Index 0 must be the BEST deal (becomes the wide featured card)
- Include a mix: fashion, home, food, tech, pets, beauty, grocery, kids
- Use real current deals you know about for ${today}
- Prioritize deals with actual promo codes`;

  const body = JSON.stringify({
    model: 'claude-sonnet-4-6',
    max_tokens: 4000,
    messages: [{ role: 'user', content: prompt }]
  });

  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: 'api.anthropic.com',
      path: '/v1/messages',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
        'x-api-key': KEY,
        'anthropic-version': '2023-06-01'
      }
    }, res => {
      let d = '';
      res.on('data', c => d += c);
      res.on('end', () => {
        try {
          const parsed = JSON.parse(d);
          if (parsed.error) throw new Error(parsed.error.message);
          const text = parsed.content?.[0]?.text || '';
          console.log('Claude response length:', text.length);
          const match = text.replace(/```json|```/g, '').trim().match(/\[[\s\S]*\]/);
          if (!match) throw new Error('No JSON array in response. Response: ' + text.slice(0, 200));
          const deals = JSON.parse(match[0]);
          if (!Array.isArray(deals) || deals.length < 8) throw new Error(`Only ${deals.length} deals returned`);
          resolve(deals);
        } catch(e) { reject(e); }
      });
    });
    req.on('error', reject);
    req.setTimeout(30000, () => { req.destroy(); reject(new Error('API timeout')); });
    req.write(body);
    req.end();
  });
}

function buildCard(d, i) {
  const f = i === 0;
  const logo = `<img src="https://logo.clearbit.com/${d.domain}" alt="${d.brand}" onerror="this.src='https://www.google.com/s2/favicons?sz=64&domain=${d.domain}';this.onerror=null;">`;
  const code = d.code
    ? `<div class="code-box" onclick="copyCode('${d.code}')"><svg viewBox="0 0 16 16" fill="none" stroke="currentColor" stroke-width="1.5"><rect x="5" y="5" width="8" height="9" rx="1"/><path d="M3 11V3a1 1 0 011-1h8"/></svg>${d.code}</div>`
    : `<span class="no-code">${d.noCode || '✓ No code needed'}</span>`;
  const orig = d.originalPrice ? `<div class="deal-original">${d.originalPrice}</div>` : '';
  const exp = d.expiry ? `<div class="deal-exp">${d.expiry}</div>` : '';
  return `
  <div class="deal-card${f ? ' featured' : ''}">
    <div class="deal-card-img"${f ? ' style="width:220px;flex-shrink:0;"' : ''}>${logo}<span class="deal-badge ${d.badgeStyle}">${d.badge}</span></div>
    <div class="deal-card-body">
      <div class="deal-source">${d.category}</div>
      <div class="deal-title"><a href="${d.link}" target="_blank">${d.title}</a></div>
      <div class="deal-desc">${d.description}</div>
      <div class="deal-footer"><div><div class="deal-discount">${d.discount}</div>${orig}</div>${code}</div>
      ${exp}
    </div>
  </div>`;
}

async function main() {
  if (!KEY) throw new Error('ANTHROPIC_API_KEY not set');
  console.log('Asking Claude for today\'s deals...');

  const deals = await getDeals();
  console.log(`Got ${deals.length} deals`);

  let html = fs.readFileSync('index.html', 'utf8');

  // Update date
  const today = new Date().toLocaleDateString('en-US', { month: 'long', day: 'numeric', year: 'numeric' });
  html = html.replace(/(<span class="badge">🔥 )([^<]+)(<\/span>)/, `$1${today}$3`);

  // Replace deals grid
  const START = '<div class="deals-grid">';
  const END = '</div><!-- end deals-grid -->';
  const si = html.indexOf(START);
  const ei = html.indexOf(END);
  if (si === -1 || ei === -1) throw new Error('deals-grid markers not found');

  const cards = deals.map((d, i) => buildCard(d, i)).join('\n');
  html = html.slice(0, si + START.length) + '\n' + cards + '\n\n' + html.slice(ei);

  // Verify
  const count = (html.match(/class="deal-card/g) || []).length;
  if (count < 8) throw new Error(`Only ${count} cards written — aborting`);

  fs.writeFileSync('index.html', html);
  console.log(`✅ Done! ${deals.length} deals for ${today}`);
}

main().catch(e => {
  console.error('FAILED:', e.message);
  process.exit(1);
});

File: update-deals-action.py
#!/usr/bin/env python3
"""
Daily deal updater for Scoop Deals
Asks Claude for today's best deals and updates index.html
"""
import os, json, re, urllib.request, urllib.error
from datetime import datetime

ANTHROPIC_KEY = os.environ["ANTHROPIC_API_KEY"]
GITHUB_TOKEN  = os.environ["GITHUB_TOKEN"]
GITHUB_OWNER  = os.environ.get("GITHUB_REPOSITORY","").split("/")[0]
GITHUB_REPO   = os.environ.get("GITHUB_REPOSITORY","").split("/")[-1]

def call_claude(prompt):
    body = json.dumps({
        "model": "claude-sonnet-4-6",
        "max_tokens": 4000,
        "messages": [{"role": "user", "content": prompt}]
    }).encode()
    
    req = urllib.request.Request(
        "https://api.anthropic.com/v1/messages",
        data=body,
        headers={
            "Content-Type": "application/json",
            "x-api-key": ANTHROPIC_KEY,
            "anthropic-version": "2023-06-01"
        },
        method="POST"
    )
    with urllib.request.urlopen(req, timeout=60) as r:
        return json.loads(r.read())

def get_deals():
    today = datetime.now().strftime("%A, %B %-d, %Y")
    prompt = f"""You are a deal curator for Scoop Deals. Today is {today}.

Find the 12 best current deals, promo codes and sales available right now from major retailers like Amazon, Target, Walmart, Best Buy, Nike, Ulta, Chewy, Kohl's, Nordstrom, Wayfair, Home Depot, Lowe's, Sephora, Bath & Body Works, Gap, Old Navy, Lululemon, REI, Dick's, CVS, Walgreens and others.

Return ONLY a JSON array of exactly 12 objects. Start with [ immediately. No markdown fences.

Each object needs these exact fields:
brand, domain, badge, badgeStyle, category, title, description, discount, originalPrice, code, noCode, link, expiry

badge: Hot Deal / New / Verified / Freebie / Flash Sale / Email Exclusive
badgeStyle: badge-hot / badge-new / badge-email / badge-web / badge-free

Index 0 = best featured deal. Mix: fashion, home, food, tech, pets, beauty, grocery, kids."""

    response = call_claude(prompt)
    text = response["content"][0]["text"]
    
    # Extract JSON array
    text = text.replace("```json","").replace("```","").strip()
    match = re.search(r'\[.*\]', text, re.DOTALL)
    if not match:
        raise ValueError(f"No JSON array found. Response: {text[:300]}")
    
    deals = json.loads(match.group(0))
    if len(deals) < 8:
        raise ValueError(f"Only {len(deals)} deals returned")
    return deals

def build_card(d, i):
    featured = i == 0
    logo = f'<img src="https://logo.clearbit.com/{d["domain"]}" alt="{d["brand"]}" onerror="this.src='https://www.google.com/s2/favicons?sz=64&domain={d["domain"]}';this.onerror=null;">'
    
    if d.get("code"):
        action = f'<div class="code-box" onclick="copyCode(\'{d["code"]}\'")><svg viewBox="0 0 16 16" fill="none" stroke="currentColor" stroke-width="1.5"><rect x="5" y="5" width="8" height="9" rx="1"/><path d="M3 11V3a1 1 0 011-1h8"/></svg>{d["code"]}</div>'
    else:
        action = f'<span class="no-code">{d.get("noCode", "✓ No code needed")}</span>'
    
    orig = f'<div class="deal-original">{d["originalPrice"]}</div>' if d.get("originalPrice") else ""
    exp  = f'<div class="deal-exp">{d["expiry"]}</div>' if d.get("expiry") else ""
    
    featured_class = " featured" if featured else ""
    img_style = ' style="width:220px;flex-shrink:0;"' if featured else ""
    
    return f"""
  <div class="deal-card{featured_class}">
    <div class="deal-card-img"{img_style}>{logo}<span class="deal-badge {d["badgeStyle"]}">{d["badge"]}</span></div>
    <div class="deal-card-body">
      <div class="deal-source">{d["category"]}</div>
      <div class="deal-title"><a href="{d["link"]}" target="_blank">{d["title"]}</a></div>
      <div class="deal-desc">{d["description"]}</div>
      <div class="deal-footer"><div><div class="deal-discount">{d["discount"]}</div>{orig}</div>{action}</div>
      {exp}
    </div>
  </div>"""

def main():
    print("Asking Claude for today's deals...")
    deals = get_deals()
    print(f"Got {len(deals)} deals")
    
    with open("index.html", "r") as f:
        html = f.read()
    
    today = datetime.now().strftime("%B %-d, %Y")
    html = re.sub(r'(<span class="badge">🔥 )([^<]+)(</span>)', f'\\g<1>{today}\\g<3>', html)
    
    cards = "\n".join(build_card(d, i) for i, d in enumerate(deals))
    
    start = html.find('<div class="deals-grid">')
    end = html.find('</div><!-- end deals-grid -->')
    if start == -1 or end == -1:
        raise ValueError("deals-grid markers not found")
    
    html = html[:start + len('<div class="deals-grid">')] + "\n" + cards + "\n\n" + html[end:]
    
    count = html.count('class="deal-card')
    if count < 8:
        raise ValueError(f"Only {count} cards written")
    
    with open("index.html", "w") as f:
        f.write(html)
    
    print(f"Done! {len(deals)} deals for {today}")

if __name__ == "__main__":
    main()
